### PR TITLE
NUX API

### PR DESCRIPTION
**Pull request by estrattonbailey**
Adds a generic storage interface for "new user experiences". In the app, these will behave like:
- user loads app and encounters a new experience
- we check this storage interface to see if the user has completed this NUX
  - if `completed === true` we do nothing
- we show the user the NUX
  - we can optionally store data on this NUX also using its `data` attribute
- when NUX is dismissed, we set `completed === true` and data is stored in prefs

The interface of a NUX is as follows:
- `id` a string with a max length of 64 characters, meant to be unique but human readable, this will be defined by the app
- `completed` a boolean, you get the idea
- `expiresAt` optional datetime, if expired, app will treat NUX as completed
- `data` string, max length of 300 chars (matches posts), optional stash of data, app will handle serde

### Validation
NUXs are validated before insertion, and before we even hit the prefs API. No additional properties beyond what is defined are allowed.

### Cleanup
Over time, these will accrue. In the app, we'll need to be judicious about how many NUXs are present for a given user. On a rolling basis, we will need to remove old NUXs as we add new ones. Some time after a NUX is removed from the app, we can remove old `id`s either on app load or on next write. This PR does not specify solutions to this.

### Naming
"New user experiences" as an acronym is NUXs. Not NUXes. `nuxs` does look a little weird, but for consistency I left it like that. NUX is how we've been referring to these internally, but if anyone has a better idea for a name lmk.

### CODE REVIEW

File: packages/api/src/agent.ts [hunk 1]
```diff
@@ -1374,6 +1377,89 @@ export class Agent extends XrpcClient {
...
+
+      bskyAppStatePref = bskyAppStatePref || {}
+      bskyAppStatePref.nuxs = bskyAppStatePref.nuxs || []
+
+      for (const id of ids) {
+        const index = bskyAppStatePref.nuxs.findIndex((n) => {
+          return n.id === id
+        })
+        if (index !== -1) {
+          bskyAppStatePref.nuxs.splice(index, 1)
```
[the remark this answers is not in the document]
Comment (gaearon) [reply packages/api/src/agent.ts:1448]: How about replacing this loop with

```js
bskyAppStatePref.nuxs = bskyAppStatePref.nuxs.filter(nux =>
  !ids.includes(nux.id)
)
```

? Would remove the need for mutation and also is a bit shorter.

[the remark this answers is not in the document]
Comment (estrattonbailey) [reply packages/api/src/agent.ts]: That's fair, swapped for a filter. We won't ever be running this in a loop or passing in lots of `id`s, so I think a simple filter and `[].includes` is fine perf-wise. Easy to grok.

[the remark this answers is not in the document]
Comment (estrattonbailey) [reply packages/api/src/agent.ts]: Oh lol hadn't reloaded the page, did exactly that

**Review by dholms — APPROVED** [review]
yup yup look good 👌

**Pull request merged**